service/history/workflowTaskHandlerCallbacks.go

```diff
@@ -405,6 +405,13 @@ func (handler *workflowTaskHandlerCallbacksImpl) handleWorkflowTaskCompleted(
 		return nil, serviceerror.NewNotFound("Workflow task not found.")
 	}
 
+	// It's an error if the workflow has used versioning in the past but this task has no versioning info.
+	if ms.GetWorkerVersionStamp() != nil {
+		if !request.UseVersioning || len(request.WorkerVersionStamp.GetBuildId()) == 0 {
+			return nil, serviceerror.NewInvalidArgument("Workflow using versioning must continue to use versioning.")
+		}
+	}
```

Comment:
When would this happen? It seems like an error case that's very difficult to recover from here

Reply:
It could happen if a user configures their workers wrong (turns off versioning after using it). Recovery is easy, they just fix their worker configuration and they'll start working again

Reply:
I see. That's reassuring, then I think we should do 2 things:
1) We should let them know how to resolve this error (how to fix their config)
2) This should probably be a FailedPrecondition. I'm not sure if we want to use InvalidArgument for async things like this

Reply:
It's not a failed precondition, the request is not retryable as-is.
https://grpc.github.io/grpc/core/md_doc_statuscodes.html

Reply:
Ok cool then invalid argument does look like the correct choice

Reply:
I still think we have room to add something about what variable in what config they should change. Other than that, this change LGTM

Reply:
We probably want metrics for some of these eventually. There'll be a bunch more cases in matching that indicate similar kinds of worker configuration errors. Users can get them from worker logs, but also might want metrics